ts/WoltLabSuite/Core/Ui/Message/Share/Dialog.ts

```diff
@@ -0,0 +1,170 @@
+ * @module  WoltLabSuite/Core/Ui/Message/Share/Dialog
+ */
+
+import UiDialog from "../../Dialog";
+import DomUtil from "../../../Dom/Util";
+import * as DomTraverse from "../../../Dom/Traverse";
+import * as Language from "../../../Language";
+import * as Clipboard from "../../../Clipboard";
+import * as UiNotification from "../../Notification";
+import * as StringUtil from "../../../StringUtil";
+import DomChangeListener from "../../../Dom/Change/Listener";
+import * as UiMessageShare from "../Share";
+import * as UiMessageShareProviders from "./Providers";
+
+const shareButtons = new WeakSet<HTMLElement>();
+
+/**
+ * Copies the contents of one of the share dialog's input elements to the clipboard.
+ */
+async function copy(event: Event): Promise<void> {
+  event.preventDefault();
+
+  const target = event.currentTarget as HTMLElement;
+  const input = DomTraverse.prevBySel(target, 'input[type="text"]') as HTMLInputElement;
+
+  await Clipboard.copyTextToClipboard(input.value);
+
+  UiNotification.show(Language.get("wcf.message.share.copy.success"));
+}
+
+/**
+ * Returns all of the dialog elements shown in the dialog.
+ */
+function getDialogElements(shareButton: HTMLElement): string {
+  let dialogOptions = "";
+
+  let permalink = "";
+  if (shareButton instanceof HTMLAnchorElement && shareButton.href) {
+    permalink = shareButton.href;
+  }
+
+  if (permalink) {
+    dialogOptions += getDialogElement("wcf.message.share.permalink", permalink);
+  }
+  if (shareButton.dataset.bbcode) {
+    dialogOptions += getDialogElement("wcf.message.share.permalink.bbcode", shareButton.dataset.bbcode);
+  }
+  if (permalink && shareButton.dataset.linkTitle) {
+    if (!shareButton.dataset.bbcode) {
+      dialogOptions += getDialogElement(
+        "wcf.message.share.permalink.bbcode",
+        `[url='${permalink}']${shareButton.dataset.linkTitle}[/url]`,
+      );
+    }
+
+    dialogOptions += getDialogElement(
+      "wcf.message.share.permalink.html",
+      '<a href="' + permalink + '">' + shareButton.dataset.linkTitle + "</a>",
+    );
+  }
+
+  return dialogOptions;
+}
+
+/**
+ * Returns a dialog element with the given label and input field value.
+ */
+function getDialogElement(label: string, value: string): string {
+  return `
+    <dl>
+      <dt>${Language.get(label)}</dt>
+      <dd>
+        <div class="inputAddon">
+          <input type="text" class="long" readonly value="${StringUtil.escapeHTML(value)}">
+          <a href="#" class="inputSuffix button jsTooltip shareDialogCopyButton" title="${Language.get(
+            "wcf.message.share.copy",
+          )}"><span class="icon icon16 fa-files-o pointer"></span></a>
+        </div>
+      </dd>
+    </dl>
+  `;
+}
+
+function getProviderButtons(): string {
+  let providerButtons = "";
+
+  UiMessageShareProviders.getEnabledProviders().forEach((provider) => {
+    const label = Language.get(provider.label);
+
+    providerButtons += `
```

Comment:
Mutating closed over variables should be avoided. Instead: `.map()`, returning the `<li>` and then `.join('\n')` the result.